oak/server/oak_runtime.cc

```diff
@@ -30,116 +30,113 @@
+  grpc_node_ = grpc_node.get();  // borrowed copy
+  nodes_[grpc_name] = std::move(grpc_node);
+
+  // Create the initial Application Node.
+  std::string node_name;
+  OakNode* app_node = CreateNode(config.initial_node(), &node_name);
+  if (app_node == nullptr) {
+    return grpc::Status(grpc::StatusCode::INVALID_ARGUMENT, "Failed to create initial Oak Node");
   }
+  LOG(INFO) << "Created Wasm node named {" << node_name << "}";
+
+  // Create an initial channel from gRPC pseudo-Node to Application Node.
+  // Both of the initial nodes have exactly one registered handle.
+  MessageChannel::ChannelHalves halves = MessageChannel::Create();
+  Handle grpc_handle =
+      grpc_node_->AddChannel(absl::make_unique<ChannelHalf>(std::move(halves.write)));
+  Handle app_handle = app_node->AddChannel(absl::make_unique<ChannelHalf>(std::move(halves.read)));
+  LOG(INFO) << "Created initial channel from Wasm node {" << grpc_name << "}." << grpc_handle
+            << " to {" << node_name << "}." << app_handle;
 
   return grpc::Status::OK;
 }
 
-std::string OakRuntime::NextNodeName(const std::string& contents) {
-  absl::MutexLock lock(&mu_);
-  int index = next_index_[contents]++;
-  return absl::StrCat(contents, "-", index);
+std::string OakRuntime::NextNodeName(const std::string& config) {
+  int index = next_index_[config]++;
+  return absl::StrCat(config, "-", index);
 }
 
-bool OakRuntime::CreateWasmNode(const std::string& contents, std::unique_ptr<ChannelHalf> half,
-                                std::string* node_name) {
-  if (wasm_contents_.count(contents) != 1) {
-    LOG(ERROR) << "failed to find Wasm bytecode Node contents with name " << contents;
-    return false;
+// Create (but don't start) a new Node instance.  Return a borrowed pointer to
+// the new Node (or nullptr on failure).
+OakNode* OakRuntime::CreateNode(const std::string& config, std::string* node_name) {
+  std::string name = NextNodeName(config);
+  std::unique_ptr<OakNode> node;
+
+  if (wasm_config_.count(config) > 0) {
+    LOG(INFO) << "Create Wasm node named {" << name << "}";
+    const std::string* module_bytes = wasm_config_[config].get();
+    node = WasmNode::Create(this, name, *module_bytes);
+  } else if (log_config_.count(config) > 0) {
+    LOG(INFO) << "Create log node named {" << name << "}";
+    node = absl::make_unique<LoggingNode>(name);
+  } else if (storage_config_.count(config) > 0) {
+    std::string address = *(storage_config_[config].get());
+    LOG(INFO) << "Create storage proxy node named {" << name << "} connecting to " << address;
+    node = absl::make_unique<StorageNode>(name, address);
+  } else {
+    LOG(ERROR) << "failed to find config with name " << config;
+    return nullptr;
   }
-  std::string name = NextNodeName(contents);
 
-  LOG(INFO) << "Create Wasm node named {" << name << "}";
+  OakNode* result = node.get();
+  if (node != nullptr) {
+    nodes_[name] = std::move(node);
```

Comment:
emplace_back?

Reply:
It's a `map` not a `vector` so there's no `emplace_back`, and `emplace` would need to be repeated in each arm above, as would the borrow.